File: undo_redo_using_stacks.py
# Undo/Redo System using Stack

undo_stack = []   # stores previous document states
redo_stack = []   # stores undone document states
document = ""     # current state of document


def make_change(new_text):
    global document
    undo_stack.append(document)   # save current state before changing
    document = new_text
    redo_stack.clear()            # clear redo stack after new change
    print("Change made.")


def undo_action():
    global document
    if not undo_stack:
        print("Nothing to undo.")
    else:
        redo_stack.append(document)
        document = undo_stack.pop()
        print("Undo done.")


def redo_action():
    global document
    if not redo_stack:
        print("Nothing to redo.")
    else:
        undo_stack.append(document)
        document = redo_stack.pop()
        print("Redo done.")


def display_document():
    print("Current document state:", document)


# --- Example usage ---
make_change("Hello")
make_change("Hello World")
make_change("Hello World!!!")

display_document()

undo_action()
display_document()

undo_action()
display_document()

redo_action()
display_document()

File: undo_redo_using_stacks.cpp
#include <iostream>
#include <stack>
#include <string>
using namespace std;

stack<string> undoStack;
stack<string> redoStack;
string document = "";

// Make a new change
void makeChange(string newText) {
    undoStack.push(document);
    document = newText;
    while (!redoStack.empty()) redoStack.pop(); // clear redo stack
    cout << "Change made.\n";
}

// Undo last change
void undoAction() {
    if (undoStack.empty()) {
        cout << "Nothing to undo.\n";
    } else {
        redoStack.push(document);
        document = undoStack.top();
        undoStack.pop();
        cout << "Undo done.\n";
    }
}

// Redo last undone change
void redoAction() {
    if (redoStack.empty()) {
        cout << "Nothing to redo.\n";
    } else {
        undoStack.push(document);
        document = redoStack.top();
        redoStack.pop();
        cout << "Redo done.\n";
    }
}

// Display document state
void displayDocument() {
    cout << "Current document state: " << document << endl;
}

int main() {
    makeChange("Hello");
    makeChange("Hello World");
    makeChange("Hello World!!!");

    displayDocument();

    undoAction();
    displayDocument();

    undoAction();
    displayDocument();

    redoAction();
    displayDocument();

    return 0;
}
